src/types/data/date.js

```diff
@@ -1,80 +1,168 @@
 import moment from 'moment';
 import React from 'react';
-import {Calendar, TransitionView} from 'react-date-picker';
+import DatetimePicker from 'yet-another-datetime-picker';
 import ImmutablePropTypes from 'react-immutable-proptypes';
-
-import DateField from '~/components/tetheredDateField';
+import TetheredComponent from 'react-tether';
 
 export default function(register) {
   register('date', {
     component: DateComponent,
-    toString: dateToString
+    toString: (value, options) => {
+      return datetimeToString(value, options.get('dateType'));
+    }
   });
 }
 
-const DateComponent = ({name, value, options, disabled, onChange, onBlur}) => {
-  return <DateField
-    className='form-data-input date'
-    value={formatDateValue(value)}
-    placeholder={options.get('placeholder')}
-    dateFormat={getDateFormat(options.get('dateType'))}
-    disabled={disabled}
-    updateOnDateClick={true}
-    onChange={e => onChange(valueToDate(e))}
-    onBlur={onBlur}
-  >
-    <TransitionView cancelButton={false} clearButton={false}>
-      <Calendar />
-    </TransitionView>
-  </DateField>;
-};
+class DateComponent extends React.Component {
+  constructor(props) {
+    super(props);
+    this.blurTimeout = 0;
+    this.state = createInitialState(props);
+  }
 
-DateComponent.propTypes = {
-  name: React.PropTypes.string.isRequired,
-  value: React.PropTypes.number,
-  options: ImmutablePropTypes.map.isRequired,
-  disabled: React.PropTypes.bool.isRequired,
-  onChange: React.PropTypes.func.isRequired,
-  onBlur: React.PropTypes.func.isRequired
-};
+  componentWillReceiveProps(props) {
+    this.setState({input: createInitialState(props).input});
+  }
 
-function formatDateValue(value) {
-  if (typeof value == 'string') {
-    return moment.utc(value);
-  } else if (Number.isFinite(value)) {
-    return moment(value * 1000);
-  } else {
-    return '';
+  render() {
+    const {options, disabled} = this.props;
+    return (
+      <div
+        className='form-data-input-date-wrapper'
+        onFocus={this.handleFocus}
+        onBlur={this.handleBlur}
+      >
+        <input
+          className='form-data-input date'
+          type='text'
+          value={this.state.input}
+          onChange={this.handleInputChange}
+          placeholder={options.get('placeholder')}
+          disabled={disabled}
+        />
+        {this.state.showPicker && !disabled ? this.renderPicker() : null}
+      </div>
+    );
   }
-}
 
-function getDateFormat(dateType) {
-  switch (dateType) {
-    case 'time':
-      return 'hh:mm a';
-    case 'datetime':
-      return 'YYYY-MM-DD hh:mm a';
-    case 'date':
-    default:
-      return 'YYYY-MM-DD';
+  renderPicker() {
+    const {options, value} = this.props;
+    let datetime = unixToDatetime(value);
+    if (!datetime.isValid()) datetime = moment();
```

Comment:
I've been preferring to keep variables `const` where possible, so this would either go into a function, or use a ternary operator such as:
```
const datetime = datetime.isValid() ?
  asdf :
  qwer
```